### PR TITLE
Dockerfile: Adds gosu support 

### PR DESCRIPTION
**Pull request by quaresmajose**
Currently is not possible to use a different user id as one used when the container is created.

| id -u
| # 10060
| docker run --rm -u $(id -u) ...

This will enter in the container with a different user as the default one created and configured. Now inside the container the user id from the host machine, 10060 in this case, not exist:

| # I have no name!@2e9103210ce0:/$
| whoami
| # whoami: cannot find name for user ID 10060
| # I have no name!
| id -u builder
| # 1000

To user the default user account with the same id for user and group we can use gosu with an entrypoint.
Launching docker with the environment argument UID and GID fixes this issue.

| echo "$(id -u)/$(id -g)"
| # 10060/100
| docker run --rm --env UID=$(id -u) --env GID=$(id -g) ...

And on the container we have the same:

| # builder@20f4393ec053:/$
| whoami
| # builder
| echo "$(id -u)/$(id -g)"
| # 10060/100

### CODE REVIEW

**Comment by quaresmajose** [comment]
This PR add support to run the SDK in multi-user environments

File: Dockerfile [hunk 1]
```diff
@@ -78,6 +78,13 @@ RUN echo $DEV_USER:$DEV_USER_PASSWD | chpasswd
 # Initialize development environment for $DEV_USER.
 RUN sudo -u $DEV_USER -H git config --global credential.helper 'cache --timeout=3600'
 
+COPY entrypoint /
+
+RUN chmod 0755 /entrypoint \
+ && sed "s/\$DOCKER_USER/$DEV_USER/g" -i /entrypoint
```
Comment (doanac) [inline Dockerfile:84]: why not just check in the file with `0o755` and the proper content. Then you avoid this extra layer in the container image.

Comment (quaresmajose) [reply Dockerfile:84]: I will change the source permission and drop the `chmod` call but the user name change in place with `sed` is to use the same user name as defined in `ARG DEV_USER=builder` otherwise can be forgotten when DEV_USER change.

Comment (doanac) [reply Dockerfile]: make sense, just move your change up a few lines so that you can do a single `RUN` that setup the git helper cache *and* does this.

Comment (quaresmajose) [reply Dockerfile]: joined with the `RUN` of the `chpasswd` a few lines before git helper cache.

**Review by doanac — COMMENTED** [review]
LGTM, but i can't comment on gosu or the actual shell script

**Comment by quaresmajose** [comment]
> LGTM, but i can't comment on gosu or the actual shell script

Without `--env UID=$(id -u)` it work as before

**Comment by quaresmajose** [comment]
Add user group support as well

File: entrypoint [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+
+set -e
+set -u
+
+: "${UID:=0}"
+: "${GID:=${UID}}"
+
+if [ "$#" = 0 ]; then
+        set -- "$(command -v bash 2>/dev/null || command -v sh)" -l
```
Comment (ricardosalveti) [inline entrypoint:10]: Why is this needed?

File: entrypoint [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+
```
Comment (ricardosalveti) [inline entrypoint:2]: Add License and Copyright.

File: Dockerfile [hunk 4]
```diff
@@ -72,8 +72,14 @@ RUN useradd -c $DEV_USER_NAME \
...
 
+# Add entrypoint to run gosu
+COPY entrypoint /
+ENTRYPOINT ["/entrypoint"]
+
 # Add default password for the SDK user (useful with sudo)
-RUN echo $DEV_USER:$DEV_USER_PASSWD | chpasswd
+# and replace entrypoint with dev user name
+RUN echo $DEV_USER:$DEV_USER_PASSWD | chpasswd && \
+	sed "s/\$DOCKER_USER/$DEV_USER/g" -i /entrypoint
```
Comment (ricardosalveti) [inline Dockerfile:82]: Can we set DOCKER_USER with @@DOCKER_USER@@ or similar to make it clear that we're really just replacing a string here?

Otherwise when I read just entrypoint it gives the impression that DOCKER_USER should be given as part of the environment.

File: entrypoint (hunk 2 above)
Comment (quaresmajose) [reply entrypoint:10]: This is needed to set the default command to execute. When the user don't provide any argument `"$#" = 0` this will set the default command to the full path bash with sh as fallback. The -l  makes it a login shell so it will read the   .bashrc, .bash_profile, etc.

So this will sets the positional parameters to `/usr/bin/bash -l` that replacing the current shell an the end `exec "$@"`

File: entrypoint (hunk 3 above)
Comment (quaresmajose) [reply entrypoint:2]: Added MIT License and Copyright (c) 2022 Foundries.io

File: Dockerfile (hunk 4 above)
Comment (quaresmajose) [reply Dockerfile]: Fixed, is much easier to read with the @@DOCKER_USER@@

**Comment by ricardosalveti** [comment]
@quaresmajose will let you merge if you think we're gtg with it.

**Comment by quaresmajose** [comment]
> @quaresmajose will let you merge if you think we're gtg with it.

I am using it locally it's been a while so will merge
